website/blog/2024-07-26-pyfunc-in-practice/index.md

```diff
@@ -0,0 +1,930 @@
+        for model_name in self.models.keys():
+            model_path = context.artifacts.get(model_name, None)
+            if model_path:
+                self.fitted_models[model_name] = joblib.load(model_path)
+            else:
+                # Print a warning if a model is not found in the context artifacts
+                print(
+                    f"Warning: {model_name} model not found in artifacts. Initialized but not fitted."
+                )
+
+        # Reconnect to the DuckDB database to load the strategies
+        conn = duckdb.connect(self.db_path)
+        # Fetch strategies from the DuckDB database into the strategies DataFrame
+        self.strategies = conn.execute("SELECT * FROM strategies").fetchdf()
+        # Close the database connection
+        conn.close()
+```
+
+Now that everything has been encapsulated in a PyFunc model, the lifecycle of the ensemble model closely mirrors that of a traditional MLflow model. The following diagram depicts the lifecycle of the model.
+
+![Ensemble Model Lifecycle](ensemble-model-lifecycle.png)
+
+
+## MLflow Tracking
+### Using the `fit` Method to Train Sub-Models  
+
+Once the data is preprocessed, we use the custom `fit` method to train all the sub-models in our Ensemble Model. This method applies grid search to find the best hyperparameters for each sub-model, fits them to the training data, and saves the trained models for future use.
+
+```python
+# Initialize the MLflow client
+client = mlflow.MlflowClient()
+
+# Set the experiment name for organizing runs in MLflow
+mlflow.set_experiment("Ensemble Model")
+
+# Load dataset from the provided URL
+data = pd.read_csv(
+    "https://github.com/zobi123/Machine-Learning-project-with-Python/blob/master/Housing.csv?raw=true"
+)
+
+# Separate features and target variable
+X = data.drop("price", axis=1)
+y = data["price"]
+
+# Split dataset into training and test sets
+X_train, X_test, y_train, y_test = train_test_split(
+    X, y, test_size=0.2, random_state=42
+)
+
+# Create a directory to save the models and related files
+os.makedirs("models", exist_ok=True)
+
+# Initialize and train the EnsembleModel
+ensemble_model = EnsembleModel()
+
+# Preprocess the training data using the defined feature engineering method
+X_train_processed = ensemble_model.feature_engineering(X_train)
+
+# Fit the models with the preprocessed training data and save them
+ensemble_model.fit(X_train_processed, y_train, save_path="models")
+
+# Infer the model signature using a small example from the training data
+example_input = X_train[:1]  # Use a single sample for signature inference
+example_input["strategy"] = "average"
+example_output = y_train[:1]
+signature = infer_signature(example_input, example_output)
+
+# Save the preprocessing pipeline to disk
+joblib.dump(ensemble_model.preprocessor, "models/preprocessor.pkl")
+
+# Define strategies for the ensemble model
+strategy_data = {
```

Comment:
I left a comment above about having a sample of this. It might be helpful just to show a single row strategy above that explains the required schema to aid the mental model while reading through things above.

Reply:
As mentioned above, added an example of the strategy DataFrame, including an overview of the schema. Thank you.